File: graph_valid_tree/bfs.py
#based on jz java version 1: BFS, Your submission beats 49.80% Submissions!
#Your submission beats 85.16% Submissions!
class Solution:
    '''
     * @param n an integer
     * @param edges a list of undirected edges
     * @return true if it's a valid tree, or false
    '''
    def validTree(self,n, edges):
        if n == 0:
            return False

        if len(edges) != n - 1:
            return False

        graph = self.initializeGraph(n, edges)

        #bfs
        Q = [0]
        visited = set() 
        visited.add(0)
        while Q:
            node = Q.pop(0)
            for neighbor in graph[node]:
                if neighbor in visited:
                    continue
                visited.add(neighbor)
                Q.append(neighbor)

        return len(visited) == n #makes sense, all n nodes are visited

    def initializeGraph(self,n, edges):
        graph = {}
        for i in range(n):
            graph[i] = set()

        for [u,v] in  edges:
            graph[u].add(v)
            graph[v].add(u)

        return graph

File: graph_valid_tree/bfs.cpp
//jz c++ bfs version,Your submission beats 30.80% Submissions!
//Your submission beats 44.60% Submissions!
#include <vector>
#include <unordered_set>
#include <queue>
class Solution {
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        vector<unordered_set<int>> g(n, unordered_set<int>());

        for (auto a : edges) {
            g[a[0]].insert(a[1]);
            g[a[1]].insert(a[0]);
        }
        unordered_set<int> v; //visited
        queue<int> q;
        q.push(0);
        v.insert(0);
        while (!q.empty()) {
            int t = q.front(); q.pop();
            for (auto a : g[t]) {
                if (v.find(a) != v.end()) return false;
                v.insert(a);
                q.push(a);
                g[a].erase(t);
            }
        }
        return v.size() == n;
    }
};
